File: Data Structures & Algorithms/4sum/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number[][]}
     */
    fourSum(nums, target) {
        nums.sort((a,b)=> a-b);
        const res = [];
        for(let i = 0; i<= nums.length-4;i++){
            if(i>0 && nums[i] === nums[i-1]) continue;
            for(let j = i+1; j<=nums.length-3;j++){
                if(j > i+1 && nums[j] === nums[j-1]) continue;
                let l = j+1;
                let r = nums.length-1;

                while(l<r){
                    const sum = nums[i]+nums[j]+nums[l]+nums[r];
                    if(sum === target){
                        res.push([nums[i],nums[j],nums[l],nums[r]]);
                        l++;
                        while( l<r && nums[l] === nums[l-1]) l++;
                    }
                    else if(sum>target) r--;
                    else l++;
                }

            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-1.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        def dfs(pairs,index,remaining):
            if remaining == 0:
                res.append(pairs)
                return 
            if index == len(nums) or remaining < 0:
                return
            dfs(pairs,index+1,remaining)    
            dfs(pairs + [nums[index]],index,remaining-nums[index])
        dfs([],0,target)
        return res



        